pkgs/development/tools/pnpm/fetch-deps/default.nix

```diff
@@ -36,15 +36,25 @@
         name = "${pname}-pnpm-deps";
 
         nativeBuildInputs = [
+          cacert
           jq
           moreutils
           pnpm
-          cacert
+          yq
         ];
 
         installPhase = ''
           runHook preInstall
 
+          lockfileVersion="$(yq -r .lockfileVersion pnpm-lock.yaml)"
+          if [[ ''${lockfileVersion:0:1} -gt ${lib.versions.major pnpm.version} ]]; then
+            echo "ERROR: lockfileVersion in pnpm-lock.yaml is to new for the provided pnpm version ${lib.versions.major pnpm.version}!"
+            exit 1
+          fi
+
+          # Don't force us to use an older version of pnpm. A newer version probably generates another FOD though.
+          jq --sort-keys "del(.. | .packageManager?)" package.json | sponge package.json
+
```

Comment:
In a similar vein as Python `relaxDepsHook`, this should probably be a flag that's off by default.